crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+    1996, 1089, 3273, 283,  1853, 1990, 882,  3033, 2419, 2102, 219,  855,
+    2681, 1848, 712,  682,  927,  1795, 461,  1891, 2877, 2522, 1894, 1010,
+    1414, 2009, 3296, 464,  2697, 816,  1352, 2679, 1274, 1052, 1025, 2132,
+    1573, 76,   2998, 3040, 1175, 2444, 394,  1219, 2300, 1455, 2117, 1607,
+    2443, 554,  1179, 2186, 2303, 2926, 2237, 525,  735,  863,  2768, 1230,
+    2572, 556,  3010, 2266, 1684, 1239, 780,  2954, 109,  1292, 1031, 1745,
+    2688, 3061, 992,  2596, 941,  892,  1021, 2390, 642,  1868, 2377, 1482,
+    1540, 540,  1678, 1626, 279,  314,  1173, 2573, 3096, 48,   667,  1920,
+    2229, 1041, 2606, 1692, 680,  2746, 568,  3312,
+};
+
+/*
+ * Second precomputed array from Appendix A of FIPS 203 (normalised positive),
+ * or else Python:
+ * ModRoots = [pow(17, 2*bitreverse(i) + 1, p) for i in range(128)]
+ */
+static const uint16_t kModRoots[128] = {
+    17,   3312, 2761, 568,  583,  2746, 2649, 680,  1637, 1692, 723,  2606,
+    2288, 1041, 1100, 2229, 1409, 1920, 2662, 667,  3281, 48,   233,  3096,
+    756,  2573, 2156, 1173, 3015, 314,  3050, 279,  1703, 1626, 1651, 1678,
+    2789, 540,  1789, 1540, 1847, 1482, 952,  2377, 1461, 1868, 2687, 642,
+    939,  2390, 2308, 1021, 2437, 892,  2388, 941,  733,  2596, 2337, 992,
+    268,  3061, 641,  2688, 1584, 1745, 2298, 1031, 2037, 1292, 3220, 109,
+    375,  2954, 2549, 780,  2090, 1239, 1645, 1684, 1063, 2266, 319,  3010,
+    2773, 556,  757,  2572, 2099, 1230, 561,  2768, 2466, 863,  2594, 735,
+    2804, 525,  1092, 2237, 403,  2926, 1026, 2303, 1143, 2186, 2150, 1179,
+    2775, 554,  886,  2443, 1722, 1607, 1212, 2117, 1874, 1455, 1029, 2300,
+    2110, 1219, 2935, 394,  885,  2444, 2154, 1175,
+};
+
+# define SIZELT(t1, t2)             ((sizeof(t1) < sizeof(t2)) ? -1 : 1)
+# define SIZE_UINT_LT_UINT32        SIZELT(unsigned int, uint32_t)
+# define CHECK_SIZE_UINT_GE_UINT32  ((void)sizeof(char[SIZE_UINT_LT_UINT32]))
+
+/*
+ * single_keccak hashes |inlen| bytes from |in| and writes |outlen| bytes of
+ * output to |out|. If the |md| specifies a fixed-output function, like
+ * SHA3-256, then |outlen| must be the correct length for that function.
+ */
+static __owur
+int single_keccak(uint8_t *out, size_t outlen, const uint8_t *in, size_t inlen,
+                  const EVP_MD *md)
+{
+    EVP_MD_CTX *mdctx;
+    int ret = 0;
+    unsigned int sz = (unsigned int) outlen;
+
+    mdctx = EVP_MD_CTX_new();
+    if (mdctx == NULL
+        || !EVP_DigestInit_ex(mdctx, md, NULL)
+        || !EVP_DigestUpdate(mdctx, in, inlen))
+        return 0;
+
+    if (EVP_MD_xof(md))
+        ret = EVP_DigestFinalXOF(mdctx, out, outlen);
+    else
+        ret = EVP_DigestFinal_ex(mdctx, out, &sz);
+    EVP_MD_CTX_free(mdctx);
+
+    return (ret && (size_t) sz == outlen);
+}
+
+/*
+ * FIPS 203, Section 4.1, equation (4.3): PRF_eta. Takes 32+1 input bytes, to
+ * produce the input to SamplePolyCBD_eta: FIPS 203, algorithm 8.
+ */
+static __owur
```

Comment:
guess you could use ossl_inline in some places here..

Reply:
I'd prefer to trust the compiler for this kind of thing.

At least outside of header files.

Reply:
Yes, some more inlining is possible, and I hope the compiler will do that. We could at some point consider giving explicit hints. Subject, first, to some measurements that show that it plausibly helps (though compilers keep changing...)